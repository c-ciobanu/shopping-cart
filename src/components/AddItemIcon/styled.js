import styled from "styled-components";

export const StyledInput = styled.input`
	width: 100%;
	height: 35px;
	padding: 0;
	border: 0;
	border-bottom: 1px solid limegreen;
	color: rgba(0, 0, 0, 0.87);
	outline: none;
`;

export const StyledForm = styled.form`
	display: grid;
	grid-template-columns: 1fr auto;
`;

export const StyledButton = styled.button`
	width: 35px;
	height: 35px;
	padding: 0;
	background-color: limegreen;
	color: #fff;
	border: 0;
	border-radius: 50%;
	outline: none;
	cursor: pointer;

	&:disabled {
		opacity: 0.25;
	}
`;
