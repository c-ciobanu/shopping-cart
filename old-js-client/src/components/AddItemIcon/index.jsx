import React, { useState } from "react";
import PropTypes from "prop-types";

import { ReactComponent as PlusIcon } from "assets/svg/plus.svg";
import { ReactComponent as CheckIcon } from "assets/svg/check.svg";

import { StyledForm, StyledButton, StyledInput, StyledDiv } from "./styled";

function AddItemIcon({ onSubmit, placeholder, title }) {
	const [isAddMode, setIsAddMode] = useState(false);
	const [inputValue, setInputValue] = useState("");

	const handleSubmit = (e) => {
		e.preventDefault();

		onSubmit(inputValue);

		e.target.reset();
		setIsAddMode(false);
		setInputValue("");
	};

	const handleInputChange = (e) => {
		setInputValue(e.target.value);
	};

	const handleInputBlur = (e) => {
		const { relatedTarget } = e;

		if (!relatedTarget || relatedTarget.type !== "submit") {
			setIsAddMode(false);
		}
	};

	return isAddMode ? (
		<StyledForm onSubmit={handleSubmit}>
			<StyledInput
				type="text"
				onChange={handleInputChange}
				onBlur={handleInputBlur}
				placeholder={placeholder}
				autoFocus={true}
			/>

			<StyledButton type="submit" title={title} disabled={inputValue === ""}>
				<CheckIcon />
			</StyledButton>
		</StyledForm>
	) : (
		<StyledDiv>
			<StyledButton type="button" title={title} onClick={() => setIsAddMode(true)}>
				<PlusIcon />
			</StyledButton>
		</StyledDiv>
	);
}

AddItemIcon.propTypes = {
	onSubmit: PropTypes.func.isRequired,
	placeholder: PropTypes.string,
	title: PropTypes.string
};

export default AddItemIcon;
