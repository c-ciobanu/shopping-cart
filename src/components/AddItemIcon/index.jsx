import React, { useState } from "react";
import PropTypes from "prop-types";

import { ReactComponent as PlusIcon } from "assets/svg/plus.svg";
import { ReactComponent as CheckIcon } from "assets/svg/check.svg";

import { StyledForm, StyledButton, StyledInput } from "./styled";

function AddItemIcon({ onSubmit, placeholder, title }) {
	const [isAddMode, setIsAddMode] = useState(false);
	const [inputValue, setInputValue] = useState("");

	const handleSubmit = (e) => {
		e.preventDefault();

		onSubmit(inputValue);

		e.target.reset();
		setIsAddMode(false);
		setInputValue("");
	};

	const handleInputChange = (e) => {
		setInputValue(e.target.value);
	};

	return (
		<StyledForm onSubmit={handleSubmit}>
			<div>{isAddMode ? <StyledInput type="text" onChange={handleInputChange} placeholder={placeholder} /> : null}</div>

			<StyledButton
				type={isAddMode ? "submit" : "button"}
				title={title}
				onClick={isAddMode ? null : () => setIsAddMode(true)}
				disabled={isAddMode && inputValue === ""}
			>
				{isAddMode ? <CheckIcon /> : <PlusIcon />}
			</StyledButton>
		</StyledForm>
	);
}

AddItemIcon.propTypes = {
	onSubmit: PropTypes.func.isRequired,
	placeholder: PropTypes.string,
	title: PropTypes.string
};

export default AddItemIcon;
