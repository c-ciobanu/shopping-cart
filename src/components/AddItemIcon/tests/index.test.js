import React from "react";
import { render, fireEvent } from "@testing-library/react";

import AddItemIcon from "../index";

describe("<AddItemIcon />", () => {
	it("Expect to not log errors in console", () => {
		const spy = jest.spyOn(global.console, "error");
		render(<AddItemIcon onSubmit={() => {}} />);
		expect(spy).not.toHaveBeenCalled();
	});

	it("Should show input after clicking button", () => {
		const placeholder = "Input placeholder";
		const { container, getByPlaceholderText } = render(<AddItemIcon onSubmit={() => {}} placeholder={placeholder} />);

		const button = container.querySelector("button");

		expect(button).not.toHaveAttribute("disabled");
		expect(() => getByPlaceholderText(placeholder)).toThrow();

		fireEvent.click(button);

		expect(button).toHaveAttribute("disabled");
		expect(() => getByPlaceholderText(placeholder)).not.toThrow();
	});

	it("Should call onSubmit prop on form submit with the correct value", () => {
		const onSubmitMock = jest.fn();
		const inputValue = "test";

		const { container } = render(<AddItemIcon onSubmit={onSubmitMock} />);

		const button = container.querySelector("button");

		fireEvent.click(button);

		const input = container.querySelector("input");

		fireEvent.change(input, { target: { value: inputValue } });
		fireEvent.click(button);

		expect(onSubmitMock).toHaveBeenCalledTimes(1);
		expect(onSubmitMock).toHaveBeenCalledWith(inputValue);
	});

	it("Should render and match the snapshot", () => {
		const {
			container: { firstChild }
		} = render(<AddItemIcon onSubmit={() => {}} />);
		expect(firstChild).toMatchSnapshot();
	});
});
