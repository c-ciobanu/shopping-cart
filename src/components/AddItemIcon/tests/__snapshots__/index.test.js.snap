// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddItemIcon /> Should render and match the snapshot 1`] = `
<form
  class="sc-AykKD dIDREQ"
>
  <div />
  <button
    class="sc-AykKE gTIRaH"
    type="button"
  >
    <svg>
      plus.svg
    </svg>
  </button>
</form>
`;
